RecoTracker/TkSeedGenerator/src/SeedingOTEDProducer.cc

```diff
@@ -0,0 +1,321 @@
+#include "RecoTracker/TkSeedGenerator/interface/SeedingOTEDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+
+#include "Geometry/Records/interface/TrackerTopologyRcd.h"
+
+#include "RecoTracker/Record/interface/CkfComponentsRecord.h"
+#include "RecoTracker/MeasurementDet/interface/MeasurementTracker.h"
+#include "RecoTracker/MeasurementDet/interface/MeasurementTrackerEvent.h"
+#include "TrackingTools/KalmanUpdators/interface/Chi2MeasurementEstimator.h"
+#include "TrackingTools/PatternTools/interface/TrajectoryStateUpdator.h"
+
+#include "DataFormats/TrajectoryState/interface/LocalTrajectoryParameters.h"
+#include "TrackingTools/TrajectoryState/interface/TrajectoryStateTransform.h"
+#include "TrackingTools/MeasurementDet/interface/TrajectoryMeasurementGroup.h"
+
+#include "DataFormats/SiStripDetId/interface/StripSubdetector.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+SeedingOTEDProducer::SeedingOTEDProducer(edm::ParameterSet const& conf)
+    : theUpdator(nullptr),
+      tkMeasEventToken(consumes<MeasurementTrackerEvent>(conf.getParameter<edm::InputTag>("trackerEvent"))) {
+  vhProducerToken = consumes<VectorHitCollectionNew>(edm::InputTag(conf.getParameter<edm::InputTag>("src")));
+  beamSpotToken = consumes<reco::BeamSpot>(conf.getParameter<edm::InputTag>("beamSpotLabel"));
+  updatorName = conf.getParameter<std::string>("updator");
+  vhMomHelper = new VectorHitMomentumHelper();
+  produces<TrajectorySeedCollection>();
+}
+
+SeedingOTEDProducer::~SeedingOTEDProducer() { delete vhMomHelper; }
+
+void SeedingOTEDProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("src", edm::InputTag("siPhase2VectorHits", "vectorHitsAccepted"));
+  desc.add<edm::InputTag>("trackerEvent", edm::InputTag("MeasurementTrackerEvent"));
+  desc.add<edm::InputTag>("beamSpotLabel", edm::InputTag("offlineBeamSpot"));
+  desc.add<std::string>("updator", std::string("KFUpdator"));
+  descriptions.add("SeedingOTEDProducer", desc);
+}
+
+void SeedingOTEDProducer::produce(edm::Event& event, const edm::EventSetup& es) {
+  std::unique_ptr<TrajectorySeedCollection> seedsWithVHs(new TrajectorySeedCollection());
+
+  edm::ESHandle<TrackerTopology> tTopoHandle;
+  es.get<TrackerTopologyRcd>().get(tTopoHandle);
+  tkTopo = tTopoHandle.product();
+
+  edm::ESHandle<MeasurementTracker> measurementTrackerHandle;
+  es.get<CkfComponentsRecord>().get(measurementTrackerHandle);
+  measurementTracker = measurementTrackerHandle.product();
+  edm::Handle<MeasurementTrackerEvent> measurementTrackerEvent;
+  event.getByToken(tkMeasEventToken, measurementTrackerEvent);
+
+  layerMeasurements = new LayerMeasurements(*measurementTrackerHandle, *measurementTrackerEvent);
```

Comment:
This looks like a leak. How about just
```suggestion
  LayerMeasurements layerMeasurements(*measurementTrackerHandle, *measurementTrackerEvent);
```
?

Reply:
Done

Reply:
Actually it turns out that without the `new` it segfaults at run time the first time a member function of `layerMeasurements` is accessed. I am reverting the change for now until I can understand it.

Reply:
How about at least using `std::make_unique` and `unique_ptr`?

Reply:
When I try

<layerMeasurements_ = std::make_unique<LayerMeasurements(*measurementTrackerHandle, *measurementTrackerEvent)>;>

with `std::unique_ptr<LayerMeasurements> layerMeasurements_;`

I get compiler errors:

`temporary of non-literal type 'LayerMeasurements' in a constant expression`

Reply:
The syntax should be
```cpp
layerMeasurements_ = std::make_unique<LayerMeasurements>(*measurementTrackerHandle, *measurementTrackerEvent);
```

Reply:
The `layerMeasurements_` is used only in the `produce()` so it does not have to be a member variable, no?

Reply:
Ouch, of course. Compiles now, sorry for the noise. 

`layerMeasurements_` is initialized in `produce()` and then used in `run()`. Could probably be done without making it a member variable.